### ISSUE TITLE
`apply_affine_to_boxes` does not handle flipping properly, due to the left-closed and right-open nature of the boxes

### ISSUE THREAD

**Issue by function2-llx** [Project-MONAI/MONAI#7740]
**Describe the bug**
`apply_affine_to_boxes` does not handle flipping properly, due to the left-closed and right-open nature of the boxes. That is to say, when a box is flipped, the left-closed and right-open coordinates become left-open and right closed, which should be maintained. 

**To Reproduce**
```python
import torch

from monai.apps.detection.transforms.box_ops import apply_affine_to_boxes
from monai.data import MetaTensor
import monai.transforms as mt

def main():
    for flip_axis in range(3):
        flip = mt.Flip(spatial_axis=flip_axis)
        x = torch.zeros(1, 1, 1, 1)
        flipped: MetaTensor = flip(x)
        box = torch.tensor([[0, 0, 0, 1, 1, 1]])
        box_flipped = apply_affine_to_boxes(box, flipped.affine.inverse())
        print(box_flipped)

if __name__ == '__main__':
    main()
```
**Expected behavior**
Produce the same results as `flip_boxes`. 

**Actual Results**
```
tensor([[-1,  0,  0,  0,  1,  1]])
tensor([[ 0, -1,  0,  1,  0,  1]])
tensor([[ 0,  0, -1,  1,  1,  0]])
```

**Suggestion**
My suggestion for the fix is to convert the box coordinates to be closed on the both sides before applying the affine matrix.

**Additional Context**
This is not about the major topic of this issue, but I would like to mention that in the code above I calculate the inverse of the affine matrix. It would be nice if `apply_affine_to_boxes` could do it for be internally by solving a linear equation.

**Comment by KumoLiu** [comment]
Hi @function2-llx, I'm not exactly sure what outcome you're expecting. It appears the result is correct as the six digits represent coordinates in the XYZXYZ format.

Thanks.

**Comment by function2-llx** [comment]
@KumoLiu Hello, the format is indeed correct, however, the values of coordinates are mismatched with `flip_boxes`. Let me do it again with this function:
```python
import torch

from monai.apps.detection.transforms.box_ops import apply_affine_to_boxes, flip_boxes
from monai.data import MetaTensor
import monai.transforms as mt

def main():
    for flip_axis in range(3):
        flip = mt.Flip(spatial_axis=flip_axis)
        x = torch.zeros(1, 1, 1, 1)
        flipped: MetaTensor = flip(x)
        box = torch.tensor([[0, 0, 0, 1, 1, 1]])
        box_flipped = apply_affine_to_boxes(box, flipped.affine.inverse())
        print('wrong:', box_flipped)
        box_flipped_correct = flip_boxes(box, x.shape[1:], flip_axis)
        print('correct:', box_flipped_correct)

if __name__ == '__main__':
    main()
```

The results are:
```
wrong: tensor([[-1,  0,  0,  0,  1,  1]])
correct: tensor([[0, 0, 0, 1, 1, 1]])
wrong: tensor([[ 0, -1,  0,  1,  0,  1]])
correct: tensor([[0, 0, 0, 1, 1, 1]])
wrong: tensor([[ 0,  0, -1,  1,  1,  0]])
correct: tensor([[0, 0, 0, 1, 1, 1]])
```

**Comment by KumoLiu** [comment]
Hi @function2-llx, the difference you're seeing arises from the second argument you're passing into the `flip_boxes` function, which is the `spatial_size` of the image. The `spatial_size` parameter dictates the axis along which you intend to flip. For more details, please take a look at the code [here](https://github.com/Project-MONAI/MONAI/blob/dev/monai/apps/detection/transforms/box_ops.py#L169). And if you set it to `[0, 0, 0]`, you will get the same result with the `apply_affine_to_boxes`.

This [comment](https://github.com/Project-MONAI/MONAI/issues/4027#issuecomment-1873720931) in the discussion about the issue might also provide some helpful context.

Thanks.

**Comment by function2-llx** [comment]
> The `spatial_size` parameter dictates the axis along which you intend to flip.

I'm sorry if I misunderstand anything, but isn't the `spatial_size` parameter literally indicating the size of the image that the boxes are on?

**Comment by KumoLiu** [comment]
Hi @function2-llx,

> but isn't the spatial_size parameter literally indicating the size of the image that the boxes are on?

Yes, for this `flip_boxes` and for detection. However, boxes can only be boxes; geometric data can be without reference. So the `flip_boxes` here is only used to detection application.
